File: string/889.根据前序和后序遍历构造二叉树.cpp
/*
 * @lc app=leetcode.cn id=889 lang=cpp
 *
 * [889] 根据前序和后序遍历构造二叉树
 */

#include "head.h"
using namespace std;
// @lc code=start
class Solution {
 private:
  vector<int> pre, post;

 public:
  TreeNode* build(int st, int en, int postroot) {
    if (st > en) return NULL;
    TreeNode* root = new TreeNode(pre[st]);
    if (postroot <= 0 || st == en) return root;
    int rroot = post[postroot - 1], index;
    for (int i = st; i <= en; i++) {
      if (pre[i] == rroot) {
        index = i;
        break;
      }
    }
    root->right = build(index, en, postroot - 1);
    if (index == st + 1) {
      return root;
    }
    int lroot = pre[st + 1];
    for (int i = 0; i <= postroot; i++) {
      if (post[i] == lroot) {
        lroot = i;
        break;
      }
    }
    root->left = build(st + 1, index - 1, lroot);
    return root;
  }
  TreeNode* constructFromPrePost(vector<int>& pre, vector<int>& post) {
    int n = post.size();
    this->pre.swap(pre);
    this->post.swap(post);
    TreeNode* root = build(0, n - 1, n - 1);
    return root;
  }
};

// @lc code=end

File: string/889.根据前序和后序遍历构造二叉树.go
package main

/*
 * @lc app=leetcode.cn id=889 lang=golang
 *
 * [889] 根据前序和后序遍历构造二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func build(preorder []int, pl, pr int, postorder []int, sl, sr int, index map[int]int) *TreeNode {
	if pl >= pr {
		return nil
	}
	root := &TreeNode{
		Val: preorder[pl],
	}
	if pl+1 == pr {
		return root
	}
	nextRootVal := preorder[pl+1]
	nextRootIndex := index[nextRootVal]
	leftSize := nextRootIndex - sl + 1

	root.Left = build(preorder, pl+1, pl+leftSize+1, postorder, sl, nextRootIndex+1, index)
	root.Right = build(preorder, pl+leftSize+1, pr, postorder, nextRootIndex+1, pr-1, index)
	return root
}

// preorder = [1,2,4,5,3,6,7], postorder = [4,5,2,6,7,3,1]
func constructFromPrePost(preorder []int, postorder []int) *TreeNode {
	index := make(map[int]int)
	for idx, v := range postorder {
		index[v] = idx
	}
	return build(preorder, 0, len(preorder), postorder, 0, len(postorder), index)
}

// @lc code=end
